File: app/src/main/java/br/com/vicentec12/mygames/ui/add_game/AddGameViewModel.java
package br.com.vicentec12.mygames.ui.add_game;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

import br.com.vicentec12.mygames.data.model.Console;
import br.com.vicentec12.mygames.data.model.Game;
import br.com.vicentec12.mygames.data.source.console.ConsoleDataSource;
import br.com.vicentec12.mygames.data.source.console.ConsoleRepository;
import br.com.vicentec12.mygames.data.source.game.GameRepository;
import br.com.vicentec12.mygames.extensions.Event;
import br.com.vicentec12.mygames.interfaces.Callbacks;

public class AddGameViewModel extends ViewModel {

    private final MutableLiveData<Game> _game = new MutableLiveData<>();
    private final MutableLiveData<List<Console>> _consoles = new MutableLiveData<>();
    private final MutableLiveData<Event<Integer>> _message = new MutableLiveData<>();
    private final MutableLiveData<Event<Boolean>> _eventDatabase = new MutableLiveData<>();

    private final GameRepository mGameRepository;
    private final ConsoleRepository mConsoleRepository;


    public AddGameViewModel(GameRepository mGameRepository, ConsoleRepository mConsoleRepository) {
        this.mGameRepository = mGameRepository;
        this.mConsoleRepository = mConsoleRepository;
    }

    public void databaseEvent() {
        Game game = getGame().getValue();
        if (game != null) {
            if (game.getId() > 0)
                updateGame(game);
            else
                insertGame(game);
        }
    }

    public void listConsoles(Game mGame) {
        if (_consoles.getValue() == null) {
            mConsoleRepository.listConsoles(new ConsoleDataSource.OnConsolesListedCallback() {
                @Override
                public void onSucess(int message, List<Console> consoles) {
                    _consoles.setValue(consoles);
                    setGame(mGame);
                }

                @Override
                public void onErro(int message) {

                }
            });
        }
    }

    private void insertGame(Game game) {
        mGameRepository.insert(game, new Callbacks.OnLocalCallback() {
            @Override
            public void onSuccess(int mMessage) {
                _eventDatabase.setValue(new Event<>(true));
                _message.setValue(new Event<>(mMessage));
            }

            @Override
            public void onFailure(int mMessage) {
                _message.setValue(new Event<>(mMessage));
            }
        });
    }

    private void updateGame(Game game) {
        mGameRepository.update(game, new Callbacks.OnLocalCallback() {
            @Override
            public void onSuccess(int mMessage) {
                _eventDatabase.setValue(new Event<>(true));
                _message.setValue(new Event<>(mMessage));
            }

            @Override
            public void onFailure(int mMessage) {
                _message.setValue(new Event<>(mMessage));
            }
        });
    }

    public boolean hasInsert() {
        if (_game.getValue() != null)
            return _game.getValue().getId() == 0;
        return false;
    }

    public void setGame(Game game) {
        if (_game.getValue() == null)
            _game.postValue(game != null ? game : new Game());
    }

    public MutableLiveData<Game> getGame() {
        return _game;
    }

    public MutableLiveData<List<Console>> getConsoles() {
        return _consoles;
    }

    public MutableLiveData<Event<Boolean>> getEventDatabase() {
        return _eventDatabase;
    }

    public MutableLiveData<Event<Integer>> getMessage() {
        return _message;
    }

}

File: app/src/main/java/br/com/vicentec12/mygames/ui/add_game/AddGameViewModel.kt
package br.com.vicentec12.mygames.ui.add_game

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.com.vicentec12.mygames.domain.model.Console
import br.com.vicentec12.mygames.domain.model.Game
import br.com.vicentec12.mygames.domain.use_case.console.ListConsolesUseCase
import br.com.vicentec12.mygames.domain.use_case.game.InsertGameUseCase
import br.com.vicentec12.mygames.domain.use_case.game.UpdateGameUseCase
import br.com.vicentec12.mygames.extensions.error
import br.com.vicentec12.mygames.extensions.success
import br.com.vicentec12.mygames.util.Event
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AddGameViewModel @Inject constructor(
    private val mListConsolesUseCase: ListConsolesUseCase,
    private val mInsertGameUseCase: InsertGameUseCase,
    private val mUpdateGameUseCase: UpdateGameUseCase
) : ViewModel() {

    private val _game = MutableLiveData<Game>()
    val game: LiveData<Game> = _game

    private val _consoles = MutableLiveData<List<Console>>()
    val consoles: LiveData<List<Console>> = _consoles

    private val _message = MutableLiveData<Event<Int>>()
    val message: LiveData<Event<Int>> = _message

    private val _success = MutableLiveData<Event<Boolean>>()
    val success: LiveData<Event<Boolean>> = _success

    private fun insertGame(mGame: Game) {
        viewModelScope.launch {
            mInsertGameUseCase(mGame).error { mResult ->
                _message.value = Event(mResult.message)
            }.success { mResult ->
                _message.value = Event(mResult.message)
                _success.value = Event(true)
                _game.value = Game()
            }
        }
    }

    private fun updateGame(mGame: Game) {
        viewModelScope.launch {
            mUpdateGameUseCase(mGame).error { mResult ->
                _message.value = Event(mResult.message)
            }.success { mResult ->
                _message.value = Event(mResult.message)
                _success.value = Event(true)
            }
        }
    }

    fun listConsoles() {
        viewModelScope.launch {
            if (_consoles.value == null) {
                mListConsolesUseCase().error { mResult ->
                    _message.value = Event(mResult.message)
                }.success { mResult ->
                    _consoles.value = mResult.data.orEmpty()
                }
            }
        }
    }

    fun databaseEvent() {
        _game.value?.run { if (id > 0) updateGame(this) else insertGame(this) }
    }

    fun setGame(mGame: Game) {
        _game.postValue(mGame)
    }

}